<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>DSCoV Single Cell</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="book_files/libs/clipboard/clipboard.min.js"></script>
<script src="book_files/libs/quarto-html/quarto.js"></script>
<script src="book_files/libs/quarto-html/popper.min.js"></script>
<script src="book_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="book_files/libs/quarto-html/anchor.min.js"></script>
<link href="book_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="book_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="book_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="book_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="book_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">DSCoV Single Cell</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<section id="scrnaseq-analysis-with-seurat" class="level1">
<h1>scRNAseq analysis with Seurat</h1>
<section id="to-use-this-notebook" class="level2">
<h2 class="anchored" data-anchor-id="to-use-this-notebook">To use this notebook</h2>
<ol type="1">
<li>Go to ood.ccv.brown.edu (you will need an Oscar account).</li>
<li>Go to ‘Clusters’ in the blue menu bar at the top and click the drop-down that says ‘&gt;_OSCAR Shell Access’</li>
<li>Go to your home folder and make a folder called <code>dscov_singlecell_2024</code> (<code>cd ~</code> and <code>mkdir scrna_r_workshop_2023</code>), then type <code>pwd</code> to get the path to this folder.</li>
<li>Look under <code>Interactive Apps</code> in the blue menu bar and click on <code>RStudio on Singularity</code> under <code>Expert GUIs</code>.</li>
</ol>
<p>Fill in the fields as follows:</p>
<ul>
<li><code>Account</code>: leave blank<br>
</li>
<li><code>Partition</code>: leave blank<br>
</li>
<li><code>Number of hours</code>: 3<br>
</li>
<li><code>Num Cores</code>: 8<br>
</li>
<li><code>Memory</code>: 90<br>
</li>
<li><code>Singularity Container Path</code>: /gpfs/data/shared/databases/workshops/bootcamp_2023/scrna_r_workshop.sif<br>
</li>
<li><code>Package install Path</code>: leave blank<br>
</li>
<li><code>Path for R Executable</code>: This should be the full path to the <code>scrna_r_workshop_2023</code> you made in step 3.<br>
</li>
<li><code>R Module</code>: leave blank<br>
</li>
<li><code>Additional Data Path</code>: /gpfs/data/shared/databases/workshops/bootcamp_2023/scrna_r_workshop<br>
</li>
</ul>
<p>Once your job starts, click the button to connect to session.<br>
At the top of the screen you’ll see a menu bar that starts with ‘file’, click on ‘file’ and ‘open file’.<br>
Open the Rproj file.<br>
</p>
</section>
<section id="introduction-to-scrna-seq" class="level2">
<h2 class="anchored" data-anchor-id="introduction-to-scrna-seq">Introduction to scRNA-seq</h2>
<p>Much of this notebook is adapted from the Seurat vignettes https://satijalab.org/seurat and GitHub repository https://github.com/satijalab/seurat</p>
<p>How does scRNAseq differ from bulk RNA-seq? In bulk RNA-seq you are taking a snapshot of expression of all the cells in a sample and your measurements are aggregated across all of those cells. In scRNAseq, you can get a sense of the heterogeneity of the cells in your sample. Are there novel or rare cell types? What about cell type specific gene expression? Does the distribution of different cell types change across time or treatment? This increased resolution comes with some unique challenges:</p>
<pre><code>-   Dropouts - genes that are not detected in some cells, can lead to sparse expression matrices with many zero values.

-   Doublets - sequencing two cells at the same time and can't distinguish their expression or cell types, need to filter these out during QC.

-   Dying cells - you will lose some cells because they are dead or dying, you can also filter these out during sample QC.

-   You also should be cautious when thinking about your sample sizes. For example, you may be sequencing thousands of cells but if they all come from the same mouse you lose the ability to generalize your findings.</code></pre>
</section>
<section id="seurat-objects-overview" class="level2">
<h2 class="anchored" data-anchor-id="seurat-objects-overview">Seurat objects overview</h2>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Important
</div>
</div>
<div class="callout-body-container callout-body">
<p>In November of 2023, Seurat made a major upgrade to Seurat v5 (https://github.com/satijalab/seurat/releases), which included many new functions and other changes (https://satijalab.org/seurat/articles/announcements.html#changes-in-seurat-v5), including some very big changes to the default behavior of Seurat. <strong>You will likely see different results depending on which version of Seurat you have used for your analysis.</strong> Feel free to come to our office hours if you want help setting up reproducible analyses using either version of Seurat.</p>
</div>
</div>
<p>This workshop focuses on using Seurat objects to structure your scRNA-seq data (https://github.com/satijalab/seurat/wiki/Seurat), we will attempt to cover how to interact with Seurat objects in Seurat v4 and v5, but won’t exhaustively cover the differences between the two versions.</p>
<p>Here’s a schematic of a Seurat object:</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="image/seurat_object.png" width="470" height="400" class="figure-img"></p>
<figcaption class="figure-caption">Schematic of a Seurat object</figcaption>
</figure>
</div>
<ul>
<li>Each Seurat object is composed of different components:
<ul>
<li><strong><code>assays</code></strong> is a list of all the assays in the object.
<ul>
<li><p>Defaults to <code>RNA</code> assay, but you can add others (like <code>SCT</code> for normalizd counts, shown in the figure above, could also be antibody-derived tags, etc.).</p></li>
<li><p>You can see all assays using <code>Assays(ifnb)</code>, see which assay is the currently active assay by looking in the <code>active.assay</code> slot (<code>ifnb@active.assay</code>) and switch between them using the <code>DefaultAssay()</code> function (<code>DefaultAssay(ifnb) &lt;- 'RNA'</code>).</p></li>
<li><p>Each assay will store multiple transformations of the data in different <code>slots</code> (or <code>layers</code> in Seurat v5) – in the case of <code>RNA</code> data these slots are:</p>
<ul>
<li><code>@counts</code> contains the raw counts.</li>
<li><code>@data</code> contains the normalized counts.</li>
<li><code>@scale.data</code> contains the scaled data for dimensional reduction.</li>
</ul></li>
<li><p>The <code>slots</code> (Seurat v4) or <code>layers</code> (Seurat v5) store the data as a sparse matrix where the rows are gene and the columns are cells.</p></li>
<li><p>In Seurat v4, you could access the raw counts like this:<code>GetAssayData(ifnb, assay="RNA", slot='counts')</code>. This will still work in Seurat v5, but you’ll get a warning message. In Seurat v5 it is intended that you access the counts using the <code>LayerData</code> function, like this: <code>LayerData(ifnb, assay='RNA', layer='counts')</code></p></li>
<li><p>In either version of Seurat <code>ifnb[['RNA']]$counts</code> will also work.</p></li>
</ul></li>
<li><strong><code>meta.data</code></strong> is a matrix of all the cell-level metadata.
<ul>
<li>This will include information about which condition, timepoint, batch, etc. a for a given cell.</li>
<li>It also includes metrics that will be relevant for QC, like <code>nCount_RNA</code> and <code>nFeature_RNA</code>
<ul>
<li><code>nCount_RNA</code> is the total number of molecules (UMIs) detected within a cell.</li>
<li><code>nFeature_RNA</code> is the total number of genes detected within a cell.</li>
</ul></li>
<li>Once you have completed clustering, you’ll also see information about which cluster each cell has been assigned to.</li>
<li>The different categories or columns in the <code>meta.data</code> are also called <code>Idents</code> in Seurat.</li>
<li>You can see the current <code>Ident</code> in the <code>active.ident</code> slot (<code>ifnb@active.ident</code>) and switch between them using the <code>Idents()</code> function (this will probably be important for running differential expression testing).</li>
<li>You can use <code>table(Idents(ifnb))</code> for a quick summary of the number of cells in each grouping.</li>
</ul></li>
<li><strong><code>graphs</code></strong> is a list of the nearest neighbor graphs.
<ul>
<li>The objects stored in <code>graphs</code> are cell x cell matrices containing the neighborhood overlap (Jaccard index) between every cell and its nearest neighbors.</li>
</ul></li>
<li><strong><code>reductions</code></strong> is a list of <code>DimReduc</code> objects.</li>
<li><strong><code>version</code></strong> contains information about which version of Seurat was used to make the object.</li>
<li>There are other optional slots, including <strong><code>tools</code></strong> and <strong><code>misc</code></strong> that can be populated by specific analysis tools (<code>tools</code>) or users can store their own additional information (<code>misc</code>).</li>
</ul></li>
</ul>
</section>
<section id="parallelization-options-for-seurat-and-other-packages" class="level2">
<h2 class="anchored" data-anchor-id="parallelization-options-for-seurat-and-other-packages">Parallelization options for Seurat and other packages</h2>
<p>First, we can set the <code>.libPaths()</code>, which essentially tells R that it should look for packages inside these locations inside the Singularity container.</p>
<div class="cell" data-layout-align="center" data-hash="book_cache/html/set lib paths_fa2d78c368a7de11f8106445785cdd43">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">.libPaths</span>(<span class="fu">c</span>(<span class="st">'/usr/local/lib/R/site-library'</span>, <span class="st">'/usr/local/lib/R/library'</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>All of the methods we are discussing here involve computationally heavy methods, and as such also take advantage of parallelization where they can. Often in their documentation you will find how to use multiple cores when calling a function, usually involving a provided a rgument or involving a package called <code>future</code>. For example, Seurat has a <a href="https://satijalab.org/seurat/archive/v4.3/future_vignette">vignette on parallelization</a> with <code>future</code>. We will follow it here:</p>
<div class="cell" data-layout-align="center" data-hash="book_cache/html/unnamed-chunk-2_7a4069db1758470b59cef87d14961c4d">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(future)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="co"># check the current active plan</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plan</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>sequential:
- args: function (..., envir = parent.frame())
- tweaked: FALSE
- call: NULL</code></pre>
</div>
</div>
<p><code>plan()</code> says that we are currently set up to run code <em>sequentially</em> or non-parallelized. To see more information, run this code chunk:</p>
<div class="cell" data-layout-align="center" data-hash="book_cache/html/unnamed-chunk-3_3ef71bf8a84e4fbf8d33ce710bf8c621">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>?future<span class="sc">::</span>plan</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now, we set workers=8 because we’ve requested 8 cores. Additionally, we set <code>multisession</code> instead of <code>multiprocess</code> despite what the vignette says, because <code>multiprocess</code> is actually deprecated in <code>future</code> and we should be explicitly specifying <code>multisession</code> or <code>multico re</code> instead. Getting into the difference is out of scope of this workshop, but you can <a href="https://cran.r-project.org/web/packages/future/vignettes/future-1-overview.html">read more</a> on future yourself if interested.</p>
<div class="cell" data-layout-align="center" data-hash="book_cache/html/unnamed-chunk-4_fa82d263ed8ad9669218025d7a60d9a0">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># change the current plan to access parallelization</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plan</span>(<span class="st">"multisession"</span>, <span class="at">workers =</span> <span class="dv">4</span>)</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plan</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>multisession:
- args: function (..., workers = 4, envir = parent.frame())
- tweaked: TRUE
- call: plan("multisession", workers = 4)</code></pre>
</div>
</div>
<p>We’ll also set a seed at the start of the notebook so that we can reproduce our results if we decide to re-run this notebook at some future date. We also set <code>future.globals.maxSize</code>, see the Seurat future vignette linked above for discussion about why we do this (basical ly we might be exceeding the allowed global variable size so we make that default bigger). We will also track how long it takes us to run through this notebook:</p>
<div class="cell" data-layout-align="center" data-hash="book_cache/html/unnamed-chunk-5_c5e637476e350dd8c488939fdd94be8d">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">61</span>)</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">future.globals.maxSize =</span> <span class="dv">4000</span> <span class="sc">*</span> <span class="dv">1024</span><span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>nb.start.time <span class="ot">&lt;-</span> <span class="fu">Sys.time</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="importing-data-and-interacting-with-seurat-objects" class="level2">
<h2 class="anchored" data-anchor-id="importing-data-and-interacting-with-seurat-objects">Importing data and interacting with Seurat objects</h2>
<p><strong>Much of this notebook is taken from the various Seurat vignettes: https://satijalab.org/seurat/articles/get_started.html</strong> First, load all the libraries we need, including some Seurat data packages. The last line will update the Seurat objects so that they are compatible with the newest version of Seurat.</p>
<div class="cell" data-layout-align="center" data-hash="book_cache/html/unnamed-chunk-6_9fd57f5cad126701593121fc57f3177d">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(RColorBrewer)</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(hdf5r)</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(stringr)</span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(biomaRt)</span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(viridis)</span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SeuratData)</span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">'ifnb.SeuratData'</span>)</span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">"ifnb"</span>)</span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a>ifnb <span class="ot">&lt;-</span> <span class="fu">UpdateSeuratObject</span>(ifnb)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>We are using the <code>SeuratData</code> package for some test data.</li>
<li>Use <code>AvailableData()</code> to see what datasets are available</li>
</ul>
<div class="cell" data-layout-align="center" data-hash="book_cache/html/unnamed-chunk-7_a33851c34ad5647780fc51a4ccf73ac9">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>SeuratData<span class="sc">::</span><span class="fu">AvailableData</span>() <span class="sc">%&gt;%</span> <span class="fu">data.frame</span>() <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                                   Dataset Version
adiposeref.SeuratData           adiposeref   1.0.0
bmcite.SeuratData                   bmcite   0.3.0
bonemarrowref.SeuratData     bonemarrowref   1.0.0
cbmc.SeuratData                       cbmc   3.1.4
celegans.embryo.SeuratData celegans.embryo   0.1.0
fetusref.SeuratData               fetusref   1.0.0
                                                                          Summary
adiposeref.SeuratData                                  Azimuth Reference: adipose
bmcite.SeuratData                                           30k Bone Marrow Cells
bonemarrowref.SeuratData                            Azimuth Reference: bonemarrow
cbmc.SeuratData                      scRNAseq and 13-antibody sequencing of CBMCs
celegans.embryo.SeuratData 6k C. elegans embryos from Packer and Zhu et al (2019)
fetusref.SeuratData                                      Azimuth Reference: fetus
                              species            system ncells
adiposeref.SeuratData           human           adipose 160075
bmcite.SeuratData               human       bone marrow  30672
bonemarrowref.SeuratData        human        bonemarrow 297627
cbmc.SeuratData                 human CBMC (cord blood)   8617
celegans.embryo.SeuratData C. elegans            embryo   6188
fetusref.SeuratData             human             fetus 377456
                                             tech seurat default.dataset
adiposeref.SeuratData      scRNA-seq and sNuc-seq   &lt;NA&gt;            &lt;NA&gt;
bmcite.SeuratData                            &lt;NA&gt;  3.2.2            &lt;NA&gt;
bonemarrowref.SeuratData                   10x v2   &lt;NA&gt;            &lt;NA&gt;
cbmc.SeuratData                          CITE-seq  3.1.4             raw
celegans.embryo.SeuratData                   &lt;NA&gt;   &lt;NA&gt;             raw
fetusref.SeuratData                          &lt;NA&gt;   &lt;NA&gt;            &lt;NA&gt;
                           disk.datasets other.datasets notes Installed
adiposeref.SeuratData               &lt;NA&gt;           &lt;NA&gt;  &lt;NA&gt;     FALSE
bmcite.SeuratData                   &lt;NA&gt;           &lt;NA&gt;  &lt;NA&gt;     FALSE
bonemarrowref.SeuratData            &lt;NA&gt;           &lt;NA&gt;  &lt;NA&gt;     FALSE
cbmc.SeuratData                processed           &lt;NA&gt;  &lt;NA&gt;     FALSE
celegans.embryo.SeuratData          &lt;NA&gt;           &lt;NA&gt;  &lt;NA&gt;     FALSE
fetusref.SeuratData                 &lt;NA&gt;           &lt;NA&gt;  &lt;NA&gt;     FALSE
                           InstalledVersion
adiposeref.SeuratData                  &lt;NA&gt;
bmcite.SeuratData                      &lt;NA&gt;
bonemarrowref.SeuratData               &lt;NA&gt;
cbmc.SeuratData                        &lt;NA&gt;
celegans.embryo.SeuratData             &lt;NA&gt;
fetusref.SeuratData                    &lt;NA&gt;</code></pre>
</div>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>SeuratData<span class="sc">::</span><span class="fu">AvailableData</span>() <span class="sc">%&gt;%</span> <span class="fu">data.frame</span>() <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(Installed <span class="sc">==</span> <span class="st">'TRUE'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                  Dataset Version                           Summary species
ifnb.SeuratData      ifnb   3.1.0 IFNB-Stimulated and Control PBMCs   human
pbmc3k.SeuratData  pbmc3k   3.1.4        3k PBMCs from 10X Genomics   human
                  system ncells   tech seurat default.dataset disk.datasets
ifnb.SeuratData     PBMC  13999 10x v1   &lt;NA&gt;             raw          &lt;NA&gt;
pbmc3k.SeuratData   PBMC   2700 10x v1  3.1.4             raw          &lt;NA&gt;
                  other.datasets notes Installed InstalledVersion
ifnb.SeuratData        processed  &lt;NA&gt;      TRUE            3.1.0
pbmc3k.SeuratData   pbmc3k.final  &lt;NA&gt;      TRUE            3.1.4</code></pre>
</div>
</div>
<ul>
<li>We’ve already installed some test data in this container. You would usually be able to install more data sets using <code>InstallData</code> but won’t have permissions to install in this container.</li>
<li>It is more likely that you are using Seurat with your own data – you can use the functions <code>Read10X</code> or <code>Read10X_h5</code> to import data.</li>
<li><code>Read10X_h5</code> works with H5 files – “Hierarchical Data Format (HDF5 or H5). H5 is a binary format that can compress and access data much more efficiently than text formats such as MEX, which is especially useful when dealing with large datasets.” https://support.10xge nomics.com/single-cell-gene-expression/software/pipelines/latest/advanced/h5_matrices.</li>
<li>You can also use <code>Read10X</code> and give a path to a folder that contains your matrix, features, and barcode tsv files.</li>
<li>After you have read in the 10X data, use it as the input to the <code>CreateSeuratObject</code> function.</li>
</ul>
<p>The <code>ifnb</code> dataset is 14,000 IFNB-Stimulated and Control PBMCs (peripheral blood mononuclear cells).</p>
<div class="cell" data-layout-align="center" data-hash="book_cache/html/unnamed-chunk-8_d0f87fc9154df8f1faf99da78016a5d5">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>?ifnb</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Currently it has a <code>RNA</code> assay:</p>
<div class="cell" data-layout-align="center" data-hash="book_cache/html/unnamed-chunk-9_ddd597c9b195c02ae8a088d55034e577">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>ifnb<span class="sc">@</span>assays</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$RNA
Assay data with 14053 features for 13999 cells
First 10 features:
 AL627309.1, RP11-206L10.2, LINC00115, NOC2L, KLHL17, PLEKHN1, HES4,
ISG15, AGRN, C1orf159 </code></pre>
</div>
</div>
<p>We can look in the <code>metadata</code> slot. Each row is a cell, and we can see which experimental group the samples came from (<code>orig.ident</code> and <code>stim</code> both tell you this information), that the cell types have already been annotated (<code>seurat_annotations</code>) and that for each cell we have information about the number of genes (<code>nFeature_RNA</code>) and molecules (<code>nCount_RNA</code>) detected.</p>
<div class="cell" data-layout-align="center" data-hash="book_cache/html/unnamed-chunk-10_3a3b6ef44a3da479e7d1d07fb604bfb9">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(ifnb<span class="sc">@</span>meta.data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                  orig.ident nCount_RNA nFeature_RNA stim seurat_annotations
AAACATACATTTCC.1 IMMUNE_CTRL       3017          877 CTRL          CD14 Mono
AAACATACCAGAAA.1 IMMUNE_CTRL       2481          713 CTRL          CD14 Mono
AAACATACCTCGCT.1 IMMUNE_CTRL       3420          850 CTRL          CD14 Mono
AAACATACCTGGTA.1 IMMUNE_CTRL       3156         1109 CTRL                pDC
AAACATACGATGAA.1 IMMUNE_CTRL       1868          634 CTRL       CD4 Memory T
AAACATACGGCATT.1 IMMUNE_CTRL       1581          557 CTRL          CD14 Mono</code></pre>
</div>
</div>
<p>We can look at the Seurat object we’ve loaded from SeuratData and see that Seurat v5 assays store data in layers. These layers can store raw, un-normalized counts (layer=‘counts’), normalized data (layer=‘data’) or z-scored/variance-stabilized data (layer=‘scale.data’).</p>
<div class="cell" data-layout-align="center" data-hash="book_cache/html/unnamed-chunk-11_2edb97075838cda70c2b9815bf112ba9">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>ifnb</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
14053 features across 13999 samples within 1 assay 
Active assay: RNA (14053 features, 0 variable features)
 2 layers present: counts, data</code></pre>
</div>
</div>
<p>We will aim to eventually integrate the different samples (<code>IMMUNE_CTRL</code> and <code>IMMUNE_STIM"</code> from <code>orig.ident</code>) together. In previous versions of Seurat, we would require the data to be represented as a list different different Seurat objects. When using Seurat v5 assays, we can instead keep all the data in one object, but simply split the layers.</p>
<div class="cell" data-layout-align="center" data-hash="book_cache/html/unnamed-chunk-12_a9b00f4b87dffec0250789d670caecd2">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>ifnb[[<span class="st">"RNA"</span>]] <span class="ot">&lt;-</span> <span class="fu">split</span>(ifnb[[<span class="st">"RNA"</span>]], <span class="at">f =</span> ifnb<span class="sc">$</span>orig.ident)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Input is a v3 assay and `split()` only works for v5 assays; converting
• to a v5 assay</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Assay RNA changing from Assay to Assay5</code></pre>
</div>
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>ifnb</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
14053 features across 13999 samples within 1 assay 
Active assay: RNA (14053 features, 0 variable features)
 4 layers present: counts.IMMUNE_CTRL, counts.IMMUNE_STIM, data.IMMUNE_CTRL, data.IMMUNE_STIM</code></pre>
</div>
</div>
<p>After splitting, there are now 4 layers (a counts and data layer for each batch). Since the data is split into layers, normalization and variable feature identification is performed for each sample independently (a consensus set of variable features is automatically identified).</p>
</section>
<section id="data-qc" class="level2">
<h2 class="anchored" data-anchor-id="data-qc">Data QC</h2>
<ul>
<li>We care about the percentage of reads that map to the mitochondrial genome because high mitochondrial reads in a cell can indicate that the cells are low-quality or dying cells. The mitochondrial QC metrics are calcualted with the <code>PercentageFeatureSet()</code> function, which calculates the percentage of counts originating from a set of features. We use the set of all genes starting with MT- as a set of mitochondrial genes – the format of the mt sequences will vary depending on which organism/genome is used…(might be ‘mt-’ for example). In this <code>ifnb</code> test dataset, there are no mitochondrial reads.</li>
</ul>
<div class="cell" data-layout-align="center" data-hash="book_cache/html/Figure out format of MT gene IDs_78d951c0a37d9c04a4ed95459859f4b5">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(ifnb) <span class="sc">%&gt;%</span> <span class="fu">grep</span>(<span class="at">pattern =</span> <span class="st">'^mt-'</span>, <span class="at">ignore.case =</span> <span class="cn">TRUE</span>, <span class="at">value =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>character(0)</code></pre>
</div>
</div>
<div class="cell" data-layout-align="center" data-hash="book_cache/html/add mt percent data_632fc982e7ffe404afb02511f73c4729">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>ifnb[[<span class="st">"percent.mt"</span>]] <span class="ot">&lt;-</span> <span class="fu">PercentageFeatureSet</span>(ifnb, <span class="at">pattern =</span> <span class="st">"^MT-"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>Before we plot, we can set the order of the object idents to whatever order we’d like:</li>
</ul>
<div class="cell" data-layout-align="center" data-hash="book_cache/html/unnamed-chunk-13_0450dd39957ee46137d11937a3828226">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(ifnb) <span class="ot">&lt;-</span> <span class="st">'orig.ident'</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a><span class="fu">levels</span>(ifnb) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"IMMUNE_CTRL"</span>, <span class="st">"IMMUNE_STIM"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ul>
<li>We can also look at plots showing the distribution of the <code>percent.mt</code>, <code>nFeature_RNA</code> and <code>nCount_RNA</code></li>
<li><code>nFeature_RNA</code> is the number of genes</li>
<li><code>nCount_RNA</code> is the number of UMIs (unique molecules – like counts)</li>
</ul>
<div class="cell" data-layout-align="center" data-hash="book_cache/html/unnamed-chunk-14_f751e92f0087b960682ca91b691d5631">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="fu">VlnPlot</span>(ifnb, <span class="at">features =</span> <span class="st">"nFeature_RNA"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="book_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="fu">VlnPlot</span>(ifnb, <span class="at">features =</span> <span class="st">"nCount_RNA"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="book_files/figure-html/unnamed-chunk-14-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="fu">VlnPlot</span>(ifnb, <span class="at">features=</span><span class="st">"percent.mt"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in SingleExIPlot(type = type, data = data[, x, drop = FALSE], idents =
idents, : All cells have the same value of percent.mt.</code></pre>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="book_files/figure-html/unnamed-chunk-14-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell" data-layout-align="center" data-hash="book_cache/html/unnamed-chunk-15_b49a227b853c5d048bf4ebb3147a52e6">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="fu">FeatureScatter</span>(ifnb, <span class="at">feature1 =</span> <span class="st">"nCount_RNA"</span>, <span class="at">feature2 =</span> <span class="st">"nFeature_RNA"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="book_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="fu">FeatureScatter</span>(ifnb, <span class="at">feature1 =</span> <span class="st">"nCount_RNA"</span>, <span class="at">feature2 =</span> <span class="st">"percent.mt"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in cor(x = data[, 1], y = data[, 2]): the standard deviation is zero</code></pre>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="book_files/figure-html/unnamed-chunk-15-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="fu">FeatureScatter</span>(ifnb, <span class="at">feature1 =</span> <span class="st">"nFeature_RNA"</span>, <span class="at">feature2 =</span> <span class="st">"percent.mt"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in cor(x = data[, 1], y = data[, 2]): the standard deviation is zero</code></pre>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="book_files/figure-html/unnamed-chunk-15-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<ul>
<li>You can also just use ggplot to make your own custom visualizations of the information in the metadata.</li>
<li>We make a separate matrix called <code>qc_data</code> and sorting it based on the <code>percent.mt</code> column.</li>
<li>Then we make our own ggplot and specify that the x and y axes should be <code>nCount_RNA</code> and <code>nFeature_RNA</code> and that the points should be colored based on <code>percent.mt</code>.</li>
<li>We use <code>scale_color_gradientn</code> to specify how the points should be colored, specifying that the limit should be between 0 and 10 and that we should <code>squish</code> anything that is out of bounds (effectively making our limits 0 and &gt;10).</li>
</ul>
<div class="cell" data-layout-align="center" data-hash="book_cache/html/unnamed-chunk-16_43cdb514a2b40bcb1c6e9b9452d52c56">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>qc_data <span class="ot">&lt;-</span> ifnb<span class="sc">@</span>meta.data[<span class="fu">c</span>(<span class="st">'orig.ident'</span>,<span class="st">'nCount_RNA'</span>,<span class="st">'nFeature_RNA'</span>,<span class="st">'percent.mt'</span>)] <span class="sc">%&gt;%</span> <span class="fu">arrange</span>(percent.mt)</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(qc_data, <span class="fu">aes</span>(<span class="at">x =</span> nCount_RNA, <span class="at">y =</span> nFeature_RNA, <span class="at">color =</span> percent.mt)) <span class="sc">+</span> </span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span> </span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_gradientn</span>(<span class="at">colors =</span> <span class="fu">rev</span>(<span class="fu">brewer.pal</span>(<span class="dv">5</span>, <span class="st">"Spectral"</span>)), <span class="at">limits =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">10</span>), <span class="at">oob =</span> (scales<span class="sc">::</span>squish)) <span class="sc">+</span></span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span>orig.ident) <span class="sc">+</span> </span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="book_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<ul>
<li>Low quality cells or empty droplets might have very few genes (<code>nFeatures</code>)</li>
<li>Dead or dying cells will also have high mitochondrial reads (<code>percent.mt</code>)</li>
<li>Doublets or multiplets will have high gene counts (<code>nFeatures</code>)</li>
<li>The total number of molecules (<code>nCount</code>) detected in a cell corresponds with the number of genes (<code>nFeatures</code>)</li>
<li>Most of the cells have less than 2000 genes and less than 7000 or so UMIs.</li>
<li>Very low mitochondrial counts from the <code>ifnb</code> data and the nFeature_RNA scatter plots look strange – perhaps this dataset was pre-filtered before being packaged into SeuratData.</li>
<li>Our goal in QC filtering is to retain as much useful information as we can, while removing doublets, empty droplets, and dead cells.</li>
<li>We will pick some thresholds for filtering based off of what we see in our data, keeping in mind that if you are doing this with your own data, your plots and thresholds will probably look a bit different.</li>
</ul>
</section>
<section id="data-filtering" class="level2">
<h2 class="anchored" data-anchor-id="data-filtering">Data Filtering</h2>
<ul>
<li>Let’s filter our data using <code>subset</code>, we’ll keep cells that have between 500 and 7000 nFeature_RNA (genes), greater than 1000 molecules, and less than 5% mitochondrial reads.</li>
</ul>
<div class="cell" data-layout-align="center" data-hash="book_cache/html/unnamed-chunk-17_0edabbe69ac71d0a26fb7168a9762c41">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>ifnb_sub <span class="ot">&lt;-</span> <span class="fu">subset</span>(ifnb, <span class="at">subset =</span> nFeature_RNA <span class="sc">&gt;</span> <span class="dv">500</span> <span class="sc">&amp;</span> nFeature_RNA <span class="sc">&lt;</span> <span class="dv">7000</span> <span class="sc">&amp;</span> percent.mt <span class="sc">&lt;</span> <span class="dv">5</span> <span class="sc">&amp;</span> nCount_RNA <span class="sc">&gt;</span> <span class="dv">1000</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Then I’ll subset the object to make is easier to work with and confirm that we have similar number of cells from each experimental group:</p>
<div class="cell" data-layout-align="center" data-hash="book_cache/html/unnamed-chunk-18_4eaa725d3befa9c63343fbf3dd538a73">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>ifnb_sub<span class="ot">&lt;-</span> <span class="fu">subset</span>(ifnb_sub, <span class="at">cells =</span> <span class="fu">sample</span>(<span class="at">x =</span> <span class="fu">rownames</span>(ifnb_sub<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">@</span>cells<span class="sc">@</span>.Data), <span class="at">size =</span> <span class="dv">5000</span>) )</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(ifnb_sub<span class="sc">@</span>meta.data<span class="sc">$</span>orig.ident)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
IMMUNE_CTRL IMMUNE_STIM 
       2306        2694 </code></pre>
</div>
</div>
</section>
<section id="normalization" class="level2">
<h2 class="anchored" data-anchor-id="normalization">Normalization</h2>
<section id="theory" class="level3">
<h3 class="anchored" data-anchor-id="theory">Theory</h3>
<p>scRNAseq data is normalized so that we can mitigate technical effects while preserving the biological signal in the data – we should be able to find the biological signal in cells irrespective of how deeply we sequenced the cell. The theory behind SCTransform (https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1874-1) is very similar to the generalized linear models (GLMs) used in bulk RNAseq analysis packages like DESeq2 and edgeR. In DESeq2 a negative binomial model is fitted to the counts and the mean and dispersion (roughly speaking how variable the observed count will be from the mean count) estimates from that model are used as the test statistics for comparison between groups. The same idea applies with SCTransform, with an additional aspect where SCTransform pools information across genes with similar abundances in order to address the higher sparsity of single cell data. We generally find that SCTransform does an excellent job alleviating variance in your data from sequencing depth alone.</p>
</section>
<section id="when-should-you-not-use-sctransform" class="level3">
<h3 class="anchored" data-anchor-id="when-should-you-not-use-sctransform">When should you not use SCTransform?</h3>
<p>The paper states: <code>As our workflow leverages all genes (or a random sub-set) for the initial regularization, we make an implicit assumption that the majority of genes in the dataset do not exhibit significant biological variation...this assumption may be overly simplistic when performing scRNA-seq on a highly heterogeneous sample, we did not observe adverse affects when applying our model to human PBMC data, or any of the other datasets we examined.</code></p>
<p>SCTransform might not work well if your data is highly heterogeneous and you expect that a high proportion of genes will exhibit significant biological variation across your samples. In this case, we would recommend the more standard workflow of <code>NormalizeData</code>, <code>FindVariableFeatures</code>, and <code>ScaleData</code>.</p>
</section>
<section id="sctransform-versions" class="level3">
<h3 class="anchored" data-anchor-id="sctransform-versions">SCTransform versions</h3>
<p>Seurat v5 run SCTransform v2 (https://satijalab.org/seurat/archive/v4.3/sctransform_v2_vignette) by default, while Seurat v4 ran SCTransform v1 by default. SCTransform v2 “improves speed and memory consumption, the stability of parameter estimates, the identification of variable features, and the the ability to perform downstream differential expression analyses.” This means you might get different results if you run Seurat v5 and re-normalize data that you have previously processed with Seurat v4. If you want to change from the default veresion of SCTransform, you can add the argument <code>vst.flavor = "v1"</code> (or <code>vst.flavor = "v2"</code>))</p>
</section>
<section id="running-sctransform" class="level3">
<h3 class="anchored" data-anchor-id="running-sctransform">Running SCTransform</h3>
<p>We will normalize using SCTransform and you might get see a warning that says ‘iteration limit reached’ when you run the function. This warning can be ignored (https://github.com/satijalab/sctransform/issues/25) because the parameter estimation generating this warning is regularized later anyway. You can use the <code>vars.to.regress</code> argument to regress out nuisance variables (like cell cycle, batch effects, or <code>percent.mt</code>). By default SCTransform will only return data for variable genes in the scale data slot – adding the <code>return.only.var.genes = FALSE</code> argument to the function call to should solve this issue (https://github.com/satijalab/seurat/issues/3553). We will also track how long it takes to run:</p>
<div class="cell" data-layout-align="center" data-hash="book_cache/html/SCTransform_483aae5a0bea19f9d63f245484de399a">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>start.time <span class="ot">&lt;-</span> <span class="fu">Sys.time</span>()</span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>ifnb_sub <span class="ot">&lt;-</span> <span class="fu">SCTransform</span>(ifnb_sub, <span class="at">vars.to.regress =</span> <span class="st">"percent.mt"</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>, <span class="at">return.only.var.genes =</span> <span class="cn">FALSE</span>)</span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a>end.time <span class="ot">&lt;-</span> <span class="fu">Sys.time</span>()</span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a>end.time <span class="sc">-</span> start.time</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Time difference of 2.4923 mins</code></pre>
</div>
</div>
<p>Run PCA and make an elbow plot</p>
<div class="cell" data-layout-align="center" data-hash="book_cache/html/unnamed-chunk-19_b9b6826c65a08919817673f2a350e1f0">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>ifnb_sub <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(ifnb_sub)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in PrepDR(object = object, features = features, verbose = verbose): The
following 20 features requested have not been scaled (running reduction without
them): NUPR1, PTGDS, IFNG, CXCL13, CCL19, CH25H, HBD, CXCL5, CSRP2, LYPD2,
CCL13, ALAS2, CCNA1, CXCL9, IL6, IL27, HRASLS2, LILRA4, MMP7, TPSAB1</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>PC_ 1 
Positive:  FTL, CCL2, TIMP1, CXCL10, CCL8, C15orf48, SOD2, APOBEC3A, FTH1, LYZ 
       CCL7, FCER1G, TYROBP, IL8, CCL3, CTSL, S100A8, LGALS3, S100A9, CD63 
       LGALS1, ANXA5, S100A4, S100A11, TYMP, CST3, IDO1, CTSB, CCL4, HLA-DRA 
Negative:  RPL3, RPS18, RPS6, RPL13, RPL21, RPL13A, RPS2, RPS4X, RPS3, RPL10 
       RPS14, RPL7, GIMAP7, CCR7, LTB, PTMA, RPL32, RPS27A, RPS19, RPS3A 
       RPL34, RPS15A, RPS27, CD3D, RPL10A, LDHB, RPS5, NPM1, RPL18A, SELL 
PC_ 2 
Positive:  CCL2, CCL8, CCL7, GIMAP7, CCL5, GNLY, FTL, S100A8, CD3D, S100A9 
       NKG7, CD7, CD2, CTSL, LTB, GIMAP5, RARRES3, RPL3, IL32, GIMAP4 
       LCK, CD3E, RPS14, PRF1, TMEM66, RPL34, SELL, CD52, RPS18, CXCL3 
Negative:  HLA-DRA, CD74, HLA-DQA1, HLA-DPB1, HLA-DPA1, HLA-DRB1, HLA-DQB1, CD83, TXN, CCR7 
       MIR155HG, SYNGR2, BIRC3, HERPUD1, CD79A, IRF8, HLA-DMA, FABP5, REL, ID3 
       GPR183, CST3, HSP90AB1, DUSP4, SERPINB1, CCL22, ID2, NME1, IDO1, PMAIP1 
PC_ 3 
Positive:  RPL13, CCR7, RPS18, LTB, CCL2, RPS6, RPL32, SELL, RPL34, RPL10 
       RPL3, FTL, RPS14, PABPC1, LDHB, RPL21, RPL13A, RPS2, RPS12, RPS3A 
       RPS4X, RPL7, RPS8, RPL18A, RPS5, RPL11, RPL9, RPS15A, RPL10A, RPS27 
Negative:  GNLY, NKG7, CCL5, GZMB, PRF1, CST7, GZMH, CLIC3, APOBEC3G, GZMA 
       KLRD1, CTSW, FGFBP2, RARRES3, CHST12, FASLG, CXCR3, C1orf21, APMAP, TNFRSF18 
       HOPX, SH2D2A, LDHA, KLRC1, ALOX5AP, ID2, FCGR3A, OASL, CD8A, CD247 
PC_ 4 
Positive:  CCL2, CCL8, FTL, CCL7, CTSL, S100A9, HSPE1, IL8, HSPD1, HSP90AB1 
       CCL4, CXCL3, CSTB, CCL3, HSPB1, LGALS3, APOBEC3B, MIR155HG, SRSF7, HSPA1A 
       S100A8, CTSB, CREM, CACYBP, NOP58, SRSF2, PLA2G7, SDS, HSPA5, NME1 
Negative:  VMO1, FCGR3A, TIMP1, MS4A7, MS4A4A, CXCL16, PLAC8, LST1, TNFSF10, FAM26F 
       GBP5, HN1, AIF1, HLA-DPA1, CD86, ATP1B3, PPM1N, FGL2, GBP1, C3AR1 
       HLA-DPB1, PILRA, WARS, COTL1, FTH1, CST3, SERPINA1, CFD, ADA, CDKN1C 
PC_ 5 
Positive:  CXCL10, FCGR3A, VMO1, MS4A7, CD69, HSPB1, CREM, CCL3, SRSF7, HSPE1 
       HSP90AB1, HSPA8, FAM26F, CACYBP, TNFSF10, MS4A4A, CCL4, CTSC, HSPD1, SOD1 
       HSPH1, NOP58, UBC, GADD45B, SRSF2, YPEL5, SAT1, DDIT4, GBP5, HSPA1A 
Negative:  HLA-DRA, CD74, LYZ, HLA-DPB1, IL8, HLA-DQA1, HLA-DRB1, S100A8, TXN, HLA-DPA1 
       NKG7, GNLY, CCL5, MARCKSL1, CST7, HLA-DQB1, GZMB, S100A9, RPL3, FTL 
       RPS4X, RPS18, GAPDH, RPL10, RPL13, LGALS1, ALOX5AP, S100A10, RPS2, GPX1 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ElbowPlot</span>(ifnb_sub)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="book_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Based on this plot, we get diminishing information returned once we get above ~10 PCs. We will use this information when we run clustering.</p>
</section>
</section>
<section id="integration" class="level2">
<h2 class="anchored" data-anchor-id="integration">Integration</h2>
<p>Seurat v5 enables streamlined integrative analysis using the IntegrateLayers function. The method currently supports five integration methods. Each of these methods performs integration in low-dimensional space, and returns a dimensional reduction (i.e.&nbsp;integrated.rpca) that aims to co-embed shared cell types across batches (samples):</p>
<pre><code>Anchor-based CCA integration (method=CCAIntegration)
Anchor-based RPCA integration (method=RPCAIntegration)
Harmony (method=HarmonyIntegration)
FastMNN (method= FastMNNIntegration)
scVI (method=scVIIntegration)</code></pre>
<p>A detailed discussion of these different methods is outside the scope of this workshop, but you can find more detail on each method in Seurat’s documentation. However, the Seurat authors state:</p>
<pre><code>By identifying shared sources of variation between datasets, CCA is well-suited for identifying anchors when cell types are conserved, but there are very substantial differences in gene expression across experiments. CCA-based integration therefore enables integrative analysis when experimental conditions or disease states introduce very strong expression shifts, or when integrating datasets across modalities and species. However, CCA-based integration may also lead to overcorrection, especially when a large proportion of cells are non-overlapping across datasets.

RPCA-based integration runs significantly faster, and also represents a more conservative approach where cells in different biological states are less likely to ‘align’ after integration. We therefore recommend RPCA during integrative analysis where:

    A substantial fraction of cells in one dataset have no matching type in the other
    Datasets originate from the same platform (i.e. multiple lanes of 10x genomics)
    There are a large number of datasets or cells to integrate (see here for more tips on integrating large datasets)</code></pre>
<p>We will run <code>CCAIntegration</code> (this was the default flavor of integration in previous versions of Seurat), <code>RPCAIntegration</code>, and <code>HarmonyIntegration</code>. Note that we are specifying that we used <code>SCT</code> normalization:</p>
<div class="cell" data-layout-align="center" data-hash="book_cache/html/unnamed-chunk-20_131a43165fe2d747ecf2fb7e90792409">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a>ifnb_sub <span class="ot">&lt;-</span> <span class="fu">IntegrateLayers</span>(</span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">object =</span> ifnb_sub, <span class="at">method =</span> CCAIntegration,</span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">orig.reduction =</span> <span class="st">"pca"</span>, <span class="at">new.reduction =</span> <span class="st">"integrated.cca"</span>, <span class="at">normalization.method =</span> <span class="st">"SCT"</span>,</span>
<span id="cb51-4"><a href="#cb51-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">verbose =</span> <span class="cn">FALSE</span></span>
<span id="cb51-5"><a href="#cb51-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center" data-hash="book_cache/html/unnamed-chunk-21_f0138f8e49bfb2e3a3f75f969389acee">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a>ifnb_sub <span class="ot">&lt;-</span> <span class="fu">IntegrateLayers</span>(</span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">object =</span> ifnb_sub, <span class="at">method =</span> RPCAIntegration,</span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">orig.reduction =</span> <span class="st">"pca"</span>, <span class="at">new.reduction =</span> <span class="st">"integrated.rpca"</span>, <span class="at">normalization.method =</span> <span class="st">"SCT"</span>,</span>
<span id="cb52-4"><a href="#cb52-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">verbose =</span> <span class="cn">FALSE</span></span>
<span id="cb52-5"><a href="#cb52-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center" data-hash="book_cache/html/unnamed-chunk-22_968d86a26c545280f51a47def5787814">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a>ifnb_sub <span class="ot">&lt;-</span> <span class="fu">IntegrateLayers</span>(</span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">object =</span> ifnb_sub, <span class="at">method =</span> HarmonyIntegration,</span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">orig.reduction =</span> <span class="st">"pca"</span>, <span class="at">new.reduction =</span> <span class="st">"harmony"</span>, <span class="at">normalization.method =</span> <span class="st">"SCT"</span>,</span>
<span id="cb53-4"><a href="#cb53-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">verbose =</span> <span class="cn">FALSE</span></span>
<span id="cb53-5"><a href="#cb53-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: HarmonyMatrix is deprecated and will be removed in the future from the
API in the future</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Warning: The parameters do_pca and npcs are deprecated. They will be ignored for this function call and please remove parameters do_pca and npcs and pass to harmony cell_embeddings directly.
This warning is displayed once per session.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Warning: The parameter tau is deprecated. It will be ignored for this function call and please remove parameter tau in future function calls. Advanced users can set value of parameter tau by using parameter .options and function harmony_options().
This warning is displayed once per session.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Warning: The parameter block.size is deprecated. It will be ignored for this function call and please remove parameter block.size in future function calls. Advanced users can set value of parameter block.size by using parameter .options and function harmony_options().
This warning is displayed once per session.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Warning: The parameter max.iter.harmony is replaced with parameter max_iter. It will be ignored for this function call and please use parameter max_iter in future function calls.
This warning is displayed once per session.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Warning: The parameter max.iter.cluster is deprecated. It will be ignored for this function call and please remove parameter max.iter.cluster in future function calls. Advanced users can set value of parameter max.iter.cluster by using parameter .options and function harmony_options().
This warning is displayed once per session.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Warning: The parameter epsilon.cluster is deprecated. It will be ignored for this function call and please remove parameter epsilon.cluster in future function calls. Advanced users can set value of parameter epsilon.cluster by using parameter .options and function harmony_options().
This warning is displayed once per session.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Warning: The parameter epsilon.harmony is deprecated. It will be ignored for this function call and please remove parameter epsilon.harmony in future function calls. If users want to control if harmony would stop early or not, use parameter early_stop. Advanced users can set value of parameter epsilon.harmony by using parameter .options and function harmony_options().
This warning is displayed once per session.</code></pre>
</div>
</div>
<p>Seurat will cluster your cells into groups of cells with similar expression patterns. The first step is <code>FindNeighbors</code>, which will construct a K-nearest neighbor (KNN) graph based on the euclidean distance in PCA space, and refine the edge weights between any two cells based on the shared overlap in their local neighborhoods (Jaccard similarity). To cluster the cells, we run <code>FindClusters</code> to apply the Louvain algorithm to iteratively group cells together, with the goal of optimizing the standard modularity function. <code>FindClusters</code> takes a <code>resolution</code> argument (defaults to a value of 0.8), which sets the granularity of the clustering, setting this parameter between 0.4-1.2 typically returns good results for single-cell datasets of around 3K cells but the resolution might increase for larger datasets. Use a value above 1 if you want a larger number of communities (clusters), and a value below 1 if you want a smaller number of communities.</p>
<div class="cell" data-layout-align="center" data-hash="book_cache/html/unnamed-chunk-23_ab8d6c58029169235bfbcb5280d129eb">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a>ifnb_sub <span class="ot">&lt;-</span> <span class="fu">FindNeighbors</span>(ifnb_sub, <span class="at">reduction =</span> <span class="st">"integrated.cca"</span>, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Computing nearest neighbor graph</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Computing SNN</code></pre>
</div>
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a>ifnb_sub <span class="ot">&lt;-</span> <span class="fu">FindClusters</span>(ifnb_sub, <span class="at">resolution =</span> .<span class="dv">8</span>, <span class="at">cluster.name =</span> <span class="st">"cca_clusters"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck

Number of nodes: 5000
Number of edges: 168525

Running Louvain algorithm...
Maximum modularity in 10 random starts: 0.8647
Number of communities: 14
Elapsed time: 0 seconds</code></pre>
</div>
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a>ifnb_sub <span class="ot">&lt;-</span> <span class="fu">FindNeighbors</span>(ifnb_sub, <span class="at">reduction =</span> <span class="st">"integrated.rpca"</span>, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Computing nearest neighbor graph
Computing SNN</code></pre>
</div>
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a>ifnb_sub <span class="ot">&lt;-</span> <span class="fu">FindClusters</span>(ifnb_sub, <span class="at">resolution =</span> .<span class="dv">8</span>, <span class="at">cluster.name =</span> <span class="st">"rpca_clusters"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck

Number of nodes: 5000
Number of edges: 167362

Running Louvain algorithm...
Maximum modularity in 10 random starts: 0.8631
Number of communities: 14
Elapsed time: 0 seconds</code></pre>
</div>
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a>ifnb_sub <span class="ot">&lt;-</span> <span class="fu">FindNeighbors</span>(ifnb_sub, <span class="at">reduction =</span> <span class="st">"harmony"</span>, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Computing nearest neighbor graph
Computing SNN</code></pre>
</div>
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a>ifnb_sub <span class="ot">&lt;-</span> <span class="fu">FindClusters</span>(ifnb_sub, <span class="at">resolution =</span> .<span class="dv">8</span>, <span class="at">cluster.name =</span> <span class="st">"harmony_clusters"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck

Number of nodes: 5000
Number of edges: 167822

Running Louvain algorithm...
Maximum modularity in 10 random starts: 0.8571
Number of communities: 12
Elapsed time: 0 seconds</code></pre>
</div>
</div>
<p>Run UMAP (Uniform Manifold Approximation and Projection) dimensional reduction technique on the unintegrated data and the different integration methods:</p>
<div class="cell" data-layout-align="center" data-hash="book_cache/html/unnamed-chunk-24_5355920feb8d357a62a664a48799e0fa">
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a>ifnb_sub <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(ifnb_sub, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, <span class="at">reduction =</span> <span class="st">"pca"</span>, <span class="at">reduction.name =</span> <span class="st">"umap.unintegrated"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric
To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'
This message will be shown once per session</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>18:50:41 UMAP embedding parameters a = 0.9922 b = 1.112</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Found more than one class "dist" in cache; using the first, from namespace 'spam'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Also defined by 'BiocGenerics'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>18:50:41 Read 5000 rows and found 10 numeric columns</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>18:50:41 Using Annoy for neighbor search, n_neighbors = 30</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Found more than one class "dist" in cache; using the first, from namespace 'spam'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Also defined by 'BiocGenerics'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>18:50:41 Building Annoy index with metric = cosine, n_trees = 50</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>0%   10   20   30   40   50   60   70   80   90   100%</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>[----|----|----|----|----|----|----|----|----|----|</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>**************************************************|
18:50:41 Writing NN index file to temp file /tmp/RtmpwLzEka/file2100ecbf2cbc
18:50:42 Searching Annoy index using 4 threads, search_k = 3000
18:50:42 Annoy recall = 100%
18:50:43 Commencing smooth kNN distance calibration using 4 threads with target n_neighbors = 30
18:50:44 Initializing from normalized Laplacian + noise (using RSpectra)
18:50:44 Commencing optimization for 500 epochs, with 198264 positive edges
18:50:50 Optimization finished</code></pre>
</div>
<div class="sourceCode cell-code" id="cb88"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a>ifnb_sub <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(ifnb_sub, <span class="at">reduction =</span> <span class="st">"integrated.cca"</span>, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, <span class="at">reduction.name =</span> <span class="st">"umap.cca"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>18:50:50 UMAP embedding parameters a = 0.9922 b = 1.112
Found more than one class "dist" in cache; using the first, from namespace 'spam'
Also defined by 'BiocGenerics'
18:50:50 Read 5000 rows and found 10 numeric columns
18:50:50 Using Annoy for neighbor search, n_neighbors = 30
Found more than one class "dist" in cache; using the first, from namespace 'spam'
Also defined by 'BiocGenerics'
18:50:50 Building Annoy index with metric = cosine, n_trees = 50
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
18:50:50 Writing NN index file to temp file /tmp/RtmpwLzEka/file2100ec24a99617
18:50:50 Searching Annoy index using 4 threads, search_k = 3000
18:50:51 Annoy recall = 100%
18:50:51 Commencing smooth kNN distance calibration using 4 threads with target n_neighbors = 30
18:50:53 Initializing from normalized Laplacian + noise (using RSpectra)
18:50:53 Commencing optimization for 500 epochs, with 200910 positive edges
18:50:58 Optimization finished</code></pre>
</div>
<div class="sourceCode cell-code" id="cb90"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a>ifnb_sub <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(ifnb_sub, <span class="at">reduction =</span> <span class="st">"integrated.rpca"</span>, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, <span class="at">reduction.name =</span> <span class="st">"umap.rpca"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>18:50:58 UMAP embedding parameters a = 0.9922 b = 1.112
Found more than one class "dist" in cache; using the first, from namespace 'spam'
Also defined by 'BiocGenerics'
18:50:58 Read 5000 rows and found 10 numeric columns
18:50:58 Using Annoy for neighbor search, n_neighbors = 30
Found more than one class "dist" in cache; using the first, from namespace 'spam'
Also defined by 'BiocGenerics'
18:50:58 Building Annoy index with metric = cosine, n_trees = 50
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
18:50:59 Writing NN index file to temp file /tmp/RtmpwLzEka/file2100ec4c077749
18:50:59 Searching Annoy index using 4 threads, search_k = 3000
18:50:59 Annoy recall = 100%
18:51:00 Commencing smooth kNN distance calibration using 4 threads with target n_neighbors = 30
18:51:01 Initializing from normalized Laplacian + noise (using RSpectra)
18:51:01 Commencing optimization for 500 epochs, with 199354 positive edges
18:51:07 Optimization finished</code></pre>
</div>
<div class="sourceCode cell-code" id="cb92"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb92-1"><a href="#cb92-1" aria-hidden="true" tabindex="-1"></a>ifnb_sub <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(ifnb_sub, <span class="at">reduction =</span> <span class="st">"harmony"</span>, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, <span class="at">reduction.name =</span> <span class="st">"umap.harmony"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>18:51:07 UMAP embedding parameters a = 0.9922 b = 1.112
Found more than one class "dist" in cache; using the first, from namespace 'spam'
Also defined by 'BiocGenerics'
18:51:07 Read 5000 rows and found 10 numeric columns
18:51:07 Using Annoy for neighbor search, n_neighbors = 30
Found more than one class "dist" in cache; using the first, from namespace 'spam'
Also defined by 'BiocGenerics'
18:51:07 Building Annoy index with metric = cosine, n_trees = 50
0%   10   20   30   40   50   60   70   80   90   100%
[----|----|----|----|----|----|----|----|----|----|
**************************************************|
18:51:07 Writing NN index file to temp file /tmp/RtmpwLzEka/file2100ec45fb0eef
18:51:08 Searching Annoy index using 4 threads, search_k = 3000
18:51:08 Annoy recall = 100%
18:51:09 Commencing smooth kNN distance calibration using 4 threads with target n_neighbors = 30
18:51:10 Initializing from normalized Laplacian + noise (using RSpectra)
18:51:10 Commencing optimization for 500 epochs, with 200718 positive edges
18:51:15 Optimization finished</code></pre>
</div>
</div>
<div class="cell" data-layout-align="center" data-hash="book_cache/html/unnamed-chunk-25_176ecc7cd3bc9f00f6dac7343b6f055b">
<div class="sourceCode cell-code" id="cb94"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb94-1"><a href="#cb94-1" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(</span>
<span id="cb94-2"><a href="#cb94-2" aria-hidden="true" tabindex="-1"></a>  ifnb_sub,</span>
<span id="cb94-3"><a href="#cb94-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">reduction =</span> <span class="st">"umap.unintegrated"</span>,</span>
<span id="cb94-4"><a href="#cb94-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">group.by =</span> <span class="st">"cca_clusters"</span>,</span>
<span id="cb94-5"><a href="#cb94-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">split.by =</span> <span class="st">"stim"</span>,  </span>
<span id="cb94-6"><a href="#cb94-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">combine =</span> <span class="cn">FALSE</span>, <span class="at">label.size =</span> <span class="dv">2</span></span>
<span id="cb94-7"><a href="#cb94-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb94-8"><a href="#cb94-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb94-9"><a href="#cb94-9" aria-hidden="true" tabindex="-1"></a>p1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[[1]]</code></pre>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="book_files/figure-html/unnamed-chunk-25-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell" data-layout-align="center" data-hash="book_cache/html/unnamed-chunk-26_227af4b5a691b487bf421473116a9b1f">
<div class="sourceCode cell-code" id="cb96"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb96-1"><a href="#cb96-1" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(</span>
<span id="cb96-2"><a href="#cb96-2" aria-hidden="true" tabindex="-1"></a>  ifnb_sub,</span>
<span id="cb96-3"><a href="#cb96-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">reduction =</span> <span class="st">"umap.rpca"</span>,</span>
<span id="cb96-4"><a href="#cb96-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">group.by =</span> <span class="st">"rpca_clusters"</span>,</span>
<span id="cb96-5"><a href="#cb96-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">split.by =</span> <span class="st">"stim"</span>,</span>
<span id="cb96-6"><a href="#cb96-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">combine =</span> <span class="cn">FALSE</span>, <span class="at">label.size =</span> <span class="dv">2</span></span>
<span id="cb96-7"><a href="#cb96-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb96-8"><a href="#cb96-8" aria-hidden="true" tabindex="-1"></a>p2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[[1]]</code></pre>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="book_files/figure-html/unnamed-chunk-26-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell" data-layout-align="center" data-hash="book_cache/html/unnamed-chunk-27_be264a651f2be2a71b890b120ceeabec">
<div class="sourceCode cell-code" id="cb98"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb98-1"><a href="#cb98-1" aria-hidden="true" tabindex="-1"></a>p3 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(</span>
<span id="cb98-2"><a href="#cb98-2" aria-hidden="true" tabindex="-1"></a>  ifnb_sub,</span>
<span id="cb98-3"><a href="#cb98-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">reduction =</span> <span class="st">"umap.harmony"</span>,</span>
<span id="cb98-4"><a href="#cb98-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">group.by =</span> <span class="st">"harmony_clusters"</span>,</span>
<span id="cb98-5"><a href="#cb98-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">split.by =</span> <span class="st">"stim"</span>,</span>
<span id="cb98-6"><a href="#cb98-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">combine =</span> <span class="cn">FALSE</span>, <span class="at">label.size =</span> <span class="dv">2</span></span>
<span id="cb98-7"><a href="#cb98-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb98-8"><a href="#cb98-8" aria-hidden="true" tabindex="-1"></a>p3</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[[1]]</code></pre>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="book_files/figure-html/unnamed-chunk-27-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell" data-layout-align="center" data-hash="book_cache/html/unnamed-chunk-28_d5c07cf377d7e88bd0cfe02bb3af4be3">
<div class="sourceCode cell-code" id="cb100"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb100-1"><a href="#cb100-1" aria-hidden="true" tabindex="-1"></a>p4 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(</span>
<span id="cb100-2"><a href="#cb100-2" aria-hidden="true" tabindex="-1"></a>  ifnb_sub,</span>
<span id="cb100-3"><a href="#cb100-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">reduction =</span> <span class="st">"umap.cca"</span>,</span>
<span id="cb100-4"><a href="#cb100-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">group.by =</span> <span class="st">"cca_clusters"</span>,</span>
<span id="cb100-5"><a href="#cb100-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">split.by =</span> <span class="st">"stim"</span>,</span>
<span id="cb100-6"><a href="#cb100-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">combine =</span> <span class="cn">FALSE</span>, <span class="at">label.size =</span> <span class="dv">2</span></span>
<span id="cb100-7"><a href="#cb100-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb100-8"><a href="#cb100-8" aria-hidden="true" tabindex="-1"></a>p4</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[[1]]</code></pre>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="book_files/figure-html/unnamed-chunk-28-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>You can manually set the colors for the clusters, like this: First, get a list of all the built in colors that R knows about and use <code>grep</code> to remove anything with <code>gray</code> or <code>grey</code> in the color name.</p>
<div class="cell" data-layout-align="center" data-hash="book_cache/html/unnamed-chunk-29_240a5d63c02fc1927670bad953ecc500">
<div class="sourceCode cell-code" id="cb102"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb102-1"><a href="#cb102-1" aria-hidden="true" tabindex="-1"></a>colors <span class="ot">&lt;-</span> grDevices<span class="sc">::</span><span class="fu">colors</span>()[<span class="fu">grep</span>(<span class="st">'gr(a|e)y|light'</span>, grDevices<span class="sc">::</span><span class="fu">colors</span>(), <span class="at">invert =</span> T)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Then pick enough colors from that list so that each <code>cca_cluster</code> has its own color</p>
<div class="cell" data-layout-align="center" data-hash="book_cache/html/unnamed-chunk-30_3e501db37ce8b5e9b8d77aaecdd5e8df">
<div class="sourceCode cell-code" id="cb103"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb103-1"><a href="#cb103-1" aria-hidden="true" tabindex="-1"></a>cca_cluster_colors <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="at">x =</span> colors, <span class="at">size =</span> ifnb_sub<span class="sc">@</span>meta.data<span class="sc">$</span>cca_clusters <span class="sc">%&gt;%</span> <span class="fu">unique</span>() <span class="sc">%&gt;%</span> <span class="fu">length</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Assign names so that each color is associated with a cluster identity:</p>
<div class="cell" data-layout-align="center" data-hash="book_cache/html/unnamed-chunk-31_7e2d2fa6ccb1e3b0b18586f583bed3db">
<div class="sourceCode cell-code" id="cb104"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb104-1"><a href="#cb104-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(cca_cluster_colors) <span class="ot">&lt;-</span>ifnb_sub<span class="sc">@</span>meta.data<span class="sc">$</span>cca_clusters <span class="sc">%&gt;%</span> <span class="fu">unique</span>() </span>
<span id="cb104-2"><a href="#cb104-2" aria-hidden="true" tabindex="-1"></a>cca_cluster_colors</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>               8               10                1                2 
          "blue"   "deepskyblue1" "darkgoldenrod1"     "palegreen2" 
               0               11                3               12 
 "darkgoldenrod" "paleturquoise2"    "darkorange2"     "mistyrose4" 
               5                6                7                4 
   "aquamarine4"  "darkturquoise"  "darkseagreen4"     "chartreuse" 
               9               13 
       "sienna4"     "indianred3" </code></pre>
</div>
</div>
<p>We can do the same and use the <code>Zissou</code> palette with <code>hcl.colors</code> (run <code>hcl.pals()</code> to see all options)</p>
<div class="cell" data-layout-align="center" data-hash="book_cache/html/unnamed-chunk-32_58253d893c563ccf0c691e130d9cf5a9">
<div class="sourceCode cell-code" id="cb106"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb106-1"><a href="#cb106-1" aria-hidden="true" tabindex="-1"></a>zissou_colors <span class="ot">&lt;-</span> <span class="fu">hcl.colors</span>( ifnb_sub<span class="sc">@</span>meta.data<span class="sc">$</span>cca_clusters <span class="sc">%&gt;%</span> <span class="fu">unique</span>() <span class="sc">%&gt;%</span> <span class="fu">length</span>(), <span class="st">"Zissou 1"</span>)</span>
<span id="cb106-2"><a href="#cb106-2" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(zissou_colors) <span class="ot">&lt;-</span>ifnb_sub<span class="sc">@</span>meta.data<span class="sc">$</span>cca_clusters <span class="sc">%&gt;%</span> <span class="fu">unique</span>() </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center" data-hash="book_cache/html/unnamed-chunk-33_38cd3538c2b3d5b935da9bc997e3066c">
<div class="sourceCode cell-code" id="cb107"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb107-1"><a href="#cb107-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(</span>
<span id="cb107-2"><a href="#cb107-2" aria-hidden="true" tabindex="-1"></a>  ifnb_sub,</span>
<span id="cb107-3"><a href="#cb107-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">reduction =</span> <span class="st">"umap.cca"</span>,</span>
<span id="cb107-4"><a href="#cb107-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">group.by =</span> <span class="st">"cca_clusters"</span>,</span>
<span id="cb107-5"><a href="#cb107-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">split.by =</span> <span class="st">"stim"</span>,</span>
<span id="cb107-6"><a href="#cb107-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">combine =</span> <span class="cn">FALSE</span>, <span class="at">label.size =</span> <span class="dv">2</span>,</span>
<span id="cb107-7"><a href="#cb107-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">cols =</span> cca_cluster_colors</span>
<span id="cb107-8"><a href="#cb107-8" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[[1]]</code></pre>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="book_files/figure-html/unnamed-chunk-33-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>We can leave the legend off, use the Zissou colors, omit the <code>split_by</code> argument and the legend:</p>
<div class="cell" data-layout-align="center" data-hash="book_cache/html/unnamed-chunk-34_1850b7fe86515b08b667a1746aa90be3">
<div class="sourceCode cell-code" id="cb109"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb109-1"><a href="#cb109-1" aria-hidden="true" tabindex="-1"></a>zissou_plot <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(</span>
<span id="cb109-2"><a href="#cb109-2" aria-hidden="true" tabindex="-1"></a>  ifnb_sub,</span>
<span id="cb109-3"><a href="#cb109-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">reduction =</span> <span class="st">"umap.cca"</span>,</span>
<span id="cb109-4"><a href="#cb109-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">group.by =</span> <span class="st">"cca_clusters"</span>,</span>
<span id="cb109-5"><a href="#cb109-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">cols =</span> zissou_colors</span>
<span id="cb109-6"><a href="#cb109-6" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span> <span class="fu">NoLegend</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Then we can label the clusters:</p>
<div class="cell" data-layout-align="center" data-hash="book_cache/html/unnamed-chunk-35_8f6ffb02e70b8f386dbd582bff51a75b">
<div class="sourceCode cell-code" id="cb110"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb110-1"><a href="#cb110-1" aria-hidden="true" tabindex="-1"></a><span class="fu">LabelClusters</span>(<span class="at">plot =</span> zissou_plot , <span class="at">id =</span> <span class="st">"cca_clusters"</span>, <span class="at">box =</span> T, <span class="at">repel =</span> T)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="book_files/figure-html/unnamed-chunk-35-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Once integrative analysis is complete, you can rejoin the layers - which collapses the individual datasets together and recreates the original counts and data layers. You will need to do this before performing any differential expression analysis. However, you can always resplit the layers in case you would like to reperform integrative analysis.</p>
<div class="cell" data-layout-align="center" data-hash="book_cache/html/unnamed-chunk-36_638497d95b698f10bff993ad647dc44f">
<div class="sourceCode cell-code" id="cb111"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb111-1"><a href="#cb111-1" aria-hidden="true" tabindex="-1"></a>ifnb_sub <span class="ot">&lt;-</span> <span class="fu">JoinLayers</span>(ifnb_sub, <span class="at">assay =</span><span class="st">'RNA'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="differential-expression-analysis" class="level2">
<h2 class="anchored" data-anchor-id="differential-expression-analysis">Differential expression analysis</h2>
<p>The bulk of Seurat’s differential expression features can be accessed through the <code>FindMarkers()</code> function. By default, Seurat performs differential expression (DE) testing based on the non-parametric Wilcoxon rank sum test. To test for DE genes between two specific groups of cells, specify the <code>ident.1</code> and <code>ident.2</code> parameters. Since we normalized using SCTransform, we have to run <code>PrepSCTFindMarkers()</code> first. Given a merged object with multiple SCT models, this function uses minimum of the median UMI (calculated using the raw UMI counts) of individual objects to reverse the individual SCT regression model using minimum of median UMI as the sequencing depth covariate. The counts slot of the SCT assay is replaced with recorrected counts and the data slot is replaced with log1p of recorrected counts. Then set the <code>DefaultAssay</code> to be the RNA assay.</p>
<div class="cell" data-layout-align="center" data-hash="book_cache/html/unnamed-chunk-37_1927d7b4bbce80a26eef7fbd354731ab">
<div class="sourceCode cell-code" id="cb112"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb112-1"><a href="#cb112-1" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(ifnb_sub) <span class="ot">&lt;-</span> <span class="st">"orig.ident"</span></span>
<span id="cb112-2"><a href="#cb112-2" aria-hidden="true" tabindex="-1"></a>ifnb_sub <span class="ot">&lt;-</span> <span class="fu">PrepSCTFindMarkers</span>(ifnb_sub)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Found 2 SCT models. Recorrecting SCT counts using minimum median counts: 1674.5</code></pre>
</div>
<div class="sourceCode cell-code" id="cb114"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb114-1"><a href="#cb114-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(ifnb_sub) <span class="ot">&lt;-</span> <span class="st">"RNA"</span></span>
<span id="cb114-2"><a href="#cb114-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb114-3"><a href="#cb114-3" aria-hidden="true" tabindex="-1"></a>stim_vs_ctrl <span class="ot">&lt;-</span> <span class="fu">FindMarkers</span>(ifnb_sub, <span class="at">ident.1 =</span> <span class="st">"IMMUNE_STIM"</span>, <span class="at">ident.2 =</span> <span class="st">"IMMUNE_CTRL"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>For a (much!) faster implementation of the Wilcoxon Rank Sum Test,
(default method for FindMarkers) please install the presto package
--------------------------------------------
install.packages('devtools')
devtools::install_github('immunogenomics/presto')
--------------------------------------------
After installation of presto, Seurat will automatically use the more 
efficient implementation (no further action necessary).
This message will be shown once per session</code></pre>
</div>
<div class="sourceCode cell-code" id="cb116"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb116-1"><a href="#cb116-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(stim_vs_ctrl <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(p_val_adj <span class="sc">&lt;</span> .<span class="dv">05</span> <span class="sc">&amp;</span> avg_log2FC <span class="sc">&gt;</span> <span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      p_val avg_log2FC pct.1 pct.2 p_val_adj
IFIT1     0   73.28255 0.901 0.028         0
IFIT3     0   44.18834 0.919 0.060         0
MX1       0   27.15754 0.898 0.099         0
IFI6      0   27.55500 0.965 0.176         0
LY6E      0   23.69293 0.926 0.168         0
IFIT2     0   33.96453 0.775 0.035         0</code></pre>
</div>
</div>
<p>The results data frame has the following columns :</p>
<pre><code>p_val : p-value (unadjusted)
avg_log2FC : log fold-change of the average expression between the two groups. Positive values indicate that the feature is more highly expressed in the first group.
pct.1 : The percentage of cells where the feature is detected in the first group
pct.2 : The percentage of cells where the feature is detected in the second group
p_val_adj : Adjusted p-value, based on Bonferroni correction using all features in the dataset.</code></pre>
<p>If the <code>ident.2</code> argument is omitted, <code>FindMarkers</code> will test for differentially expressed features between the group specified by <code>ident.1</code> and all other cells. Additionally, the parameter <code>only.pos</code> can be set to TRUE to only search for positive markers, i.e.&nbsp;features that are more highly expressed in the ident.1 group.</p>
<div class="cell" data-layout-align="center" data-hash="book_cache/html/unnamed-chunk-38_32fff991038c2c740a80000b391defae">
<div class="sourceCode cell-code" id="cb119"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb119-1"><a href="#cb119-1" aria-hidden="true" tabindex="-1"></a>stim_vs_all <span class="ot">&lt;-</span> <span class="fu">FindMarkers</span>(ifnb_sub, <span class="at">ident.1 =</span> <span class="st">"IMMUNE_STIM"</span>, <span class="at">only.pos =</span> T)</span>
<span id="cb119-2"><a href="#cb119-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(stim_vs_all <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(p_val_adj <span class="sc">&lt;</span> .<span class="dv">05</span> <span class="sc">&amp;</span> avg_log2FC <span class="sc">&gt;</span> <span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      p_val avg_log2FC pct.1 pct.2 p_val_adj
IFIT1     0   73.28255 0.901 0.028         0
IFIT3     0   44.18834 0.919 0.060         0
MX1       0   27.15754 0.898 0.099         0
IFI6      0   27.55500 0.965 0.176         0
LY6E      0   23.69293 0.926 0.168         0
IFIT2     0   33.96453 0.775 0.035         0</code></pre>
</div>
</div>
<p>Since we only have two groups (<code>IMMUNE_STIM</code> and <code>IMMUNE_CTRL</code>), the results of <code>stim_v_all</code> and <code>stim_vs_ctrl</code> are comparing the same groups and results are very similar.</p>
<p>We can switch idents to find marker genes for the clusters:</p>
<div class="cell" data-layout-align="center" data-hash="book_cache/html/unnamed-chunk-39_cf07dbf46741162e2fecc4236001aba9">
<div class="sourceCode cell-code" id="cb121"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb121-1"><a href="#cb121-1" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(ifnb_sub) <span class="ot">&lt;-</span> <span class="st">'cca_clusters'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Use <code>FindAllMarkers</code> to compare each cluster to all the other clusters. For the sake of speed, we are selecting only positive genes that are expressed in at least 90% of the cells for a given cluster:</p>
<div class="cell" data-layout-align="center" data-hash="book_cache/html/unnamed-chunk-40_16d53e2b5c93cd4714cedcaf9df8aa6d">
<div class="sourceCode cell-code" id="cb122"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb122-1"><a href="#cb122-1" aria-hidden="true" tabindex="-1"></a>cca_markers <span class="ot">&lt;-</span> <span class="fu">FindAllMarkers</span>(ifnb_sub, <span class="at">min.pct =</span> .<span class="dv">90</span>, <span class="at">only.pos=</span><span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 0</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 1</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 2</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 3</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 4</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 5</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 6</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 7</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 8</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 9</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 10</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 11</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 12</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 13</code></pre>
</div>
</div>
<p>Look at the 3 marker genes with the biggest fold change per cluster</p>
<div class="cell" data-layout-align="center" data-hash="book_cache/html/unnamed-chunk-41_810ceb467eee6e571b2bd35a07f573a8">
<div class="sourceCode cell-code" id="cb137"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb137-1"><a href="#cb137-1" aria-hidden="true" tabindex="-1"></a>top_cluster_markers <span class="ot">&lt;-</span> </span>
<span id="cb137-2"><a href="#cb137-2" aria-hidden="true" tabindex="-1"></a>  cca_markers <span class="sc">%&gt;%</span> </span>
<span id="cb137-3"><a href="#cb137-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(cluster) <span class="sc">%&gt;%</span></span>
<span id="cb137-4"><a href="#cb137-4" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(avg_log2FC <span class="sc">&gt;</span> <span class="dv">1</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb137-5"><a href="#cb137-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">top_n</span>(<span class="at">n =</span> <span class="dv">3</span>, <span class="at">wt =</span> avg_log2FC)</span>
<span id="cb137-6"><a href="#cb137-6" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(top_cluster_markers, <span class="at">n =</span> <span class="dv">37</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 39 × 7
# Groups:   cluster [14]
       p_val avg_log2FC pct.1 pct.2 p_val_adj cluster gene    
       &lt;dbl&gt;      &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt; &lt;fct&gt;   &lt;chr&gt;   
 1 4.13e-254      88.8  1     0.905 5.81e-250 0       FTL     
 2 3.61e-215      33.9  0.934 0.368 5.07e-211 0       LGALS3  
 3 6.60e-113      90.2  0.985 0.687 9.28e-109 0       SAT1    
 4 2.09e- 16       2.60 0.99  0.953 2.94e- 12 1       RPS13   
 5 3.03e- 14       2.66 0.948 0.885 4.26e- 10 1       RPL29   
 6 1.59e-148      40.5  1     0.98  2.23e-144 2       RPS18   
 7 5.32e- 98      41.7  1     0.99  7.48e- 94 2       RPS2    
 8 2.23e- 32      37.6  0.995 0.97  3.13e- 28 2       RPS9    
 9 0              43.5  0.983 0.229 0         3       CCL2    
10 6.08e-126      27.6  0.992 0.765 8.54e-122 3       VIM     
11 8.78e-117      16.7  0.996 0.804 1.23e-112 3       SH3BGRL3
12 6.60e- 51       6.53 1     0.984 9.27e- 47 4       RPL32   
13 7.20e- 31       2.16 0.983 0.9   1.01e- 26 4       RPL31   
14 7.15e- 18       5.24 0.983 0.928 1.01e- 13 4       RPL30   
15 0              16.3  0.989 0.172 0         5       FCGR3A  
16 2.56e-296      12.4  0.977 0.22  3.59e-292 5       MS4A7   
17 1.50e- 72      38.0  0.98  0.872 2.11e- 68 5       HLA-E   
18 1.65e-  9       5.72 0.904 0.743 2.31e-  5 6       RPL7A   
19 1.07e-  4       9.07 0.98  0.942 1   e+  0 6       RPL26   
20 4.15e-  4       6.65 1     0.997 1   e+  0 6       RPLP1   
21 2.17e-  5       3.72 0.989 0.929 3.04e-  1 7       RPL23A  
22 2.49e-  5      11.2  1     0.984 3.49e-  1 7       RPS19   
23 9.02e-  5      10.5  0.901 0.936 1   e+  0 7       FAU     
24 2.53e- 69      28.7  0.961 0.751 3.56e- 65 8       BTG1    
25 6.21e- 35      25.9  0.906 0.812 8.73e- 31 8       UBB     
26 6.62e- 34      72.0  0.945 0.821 9.31e- 30 8       UBC     
27 9.97e-119      18.6  1     0.496 1.40e-114 9       SOD2    
28 2.42e- 75      19.8  0.917 0.465 3.39e- 71 9       CSTB    
29 3.30e- 11      14.0  1     0.999 4.64e-  7 9       HLA-B   
30 0               9.66 0.974 0.081 0         10      NKG7    
31 2.03e- 65      23.9  0.903 0.473 2.85e- 61 11      RAN     
32 1.56e- 59      30.7  0.976 0.736 2.20e- 55 11      YBX1    
33 7.94e- 38      30.8  0.988 0.887 1.12e- 33 11      PTMA    
34 2.59e-107     279.   1     0.55  3.64e-103 12      HLA-DRA 
35 1.47e-102     188.   1     0.68  2.06e- 98 12      CD74    
36 2.90e- 97     171.   0.994 0.341 4.08e- 93 12      LYZ     
37 4.25e-288      20.5  0.941 0.017 5.97e-284 13      TSPAN13 
# ℹ 2 more rows</code></pre>
</div>
</div>
<p>Make a <code>FeaturePlot</code> to look at the expression of one of the cluster markers. It will plot the <code>data</code> slot from the default assay. We can switch the default assay to SCT first and specify that we want to use the <code>data</code> slot (log1p(counts)(:</p>
<div class="cell" data-layout-align="center" data-hash="book_cache/html/unnamed-chunk-42_0e9eb001d9f9e27c48c1a7b3407866a7">
<div class="sourceCode cell-code" id="cb139"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb139-1"><a href="#cb139-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(ifnb_sub) <span class="ot">&lt;-</span> <span class="st">"SCT"</span></span>
<span id="cb139-2"><a href="#cb139-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb139-3"><a href="#cb139-3" aria-hidden="true" tabindex="-1"></a><span class="fu">FeaturePlot</span>(ifnb_sub, <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"CD74"</span>), <span class="at">reduction =</span> <span class="st">'umap.cca'</span>, <span class="at">order =</span> T, <span class="at">slot =</span> <span class="st">'data'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="book_files/figure-html/unnamed-chunk-42-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>We can adjust the default colors and use one of the <code>viridis</code> palettes:</p>
<div class="cell" data-layout-align="center" data-hash="book_cache/html/unnamed-chunk-43_c8f7332c681d76e7f12244c376e53639">
<div class="sourceCode cell-code" id="cb140"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb140-1"><a href="#cb140-1" aria-hidden="true" tabindex="-1"></a><span class="fu">FeaturePlot</span>(ifnb_sub, <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"CD74"</span>), <span class="at">reduction =</span> <span class="st">'umap.cca'</span>, <span class="at">order =</span> T, <span class="at">slot =</span> <span class="st">'data'</span>) <span class="sc">&amp;</span> <span class="fu">scale_color_gradientn</span>(<span class="at">colors =</span> <span class="fu">turbo</span>(<span class="at">n =</span> <span class="dv">10</span>, <span class="at">direction =</span> <span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.</code></pre>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="book_files/figure-html/unnamed-chunk-43-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>We can add the cluster labels:</p>
<div class="cell" data-layout-align="center" data-hash="book_cache/html/unnamed-chunk-44_26a9be8f1b7c470d377259978c7640a8">
<div class="sourceCode cell-code" id="cb142"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb142-1"><a href="#cb142-1" aria-hidden="true" tabindex="-1"></a><span class="fu">FeaturePlot</span>(ifnb_sub, <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"CD74"</span>), <span class="at">reduction =</span> <span class="st">'umap.cca'</span>, <span class="at">order =</span> T, <span class="at">slot =</span> <span class="st">'data'</span>, <span class="at">label =</span> <span class="cn">TRUE</span>, <span class="at">repel =</span> <span class="cn">TRUE</span>) <span class="sc">&amp;</span> <span class="fu">scale_color_gradientn</span>(<span class="at">colors =</span> <span class="fu">turbo</span>(<span class="at">n =</span> <span class="dv">10</span>, <span class="at">direction =</span> <span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.</code></pre>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="book_files/figure-html/unnamed-chunk-44-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>We can use <code>RColorBrewer</code> palettes instead and specify that we want to drop the colors on the extreme ends of the <code>Spectral</code> palette:</p>
<div class="cell" data-layout-align="center" data-hash="book_cache/html/unnamed-chunk-45_2138e4bf8e85f7846bc594f48fa0150e">
<div class="sourceCode cell-code" id="cb144"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb144-1"><a href="#cb144-1" aria-hidden="true" tabindex="-1"></a><span class="fu">FeaturePlot</span>(ifnb_sub, <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"CD74"</span>), <span class="at">reduction =</span> <span class="st">'umap.cca'</span>, <span class="at">order =</span> T, <span class="at">slot =</span> <span class="st">'data'</span>, <span class="at">label =</span> <span class="cn">TRUE</span>, <span class="at">repel =</span> <span class="cn">TRUE</span>) <span class="sc">&amp;</span> <span class="fu">scale_color_gradientn</span>(<span class="at">colors =</span> <span class="fu">rev</span>(<span class="fu">brewer.pal</span>(<span class="dv">10</span>, <span class="st">'Spectral'</span>))[<span class="dv">3</span><span class="sc">:</span><span class="dv">8</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.</code></pre>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="book_files/figure-html/unnamed-chunk-45-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>And add a legend title</p>
<div class="cell" data-layout-align="center" data-hash="book_cache/html/unnamed-chunk-46_e130c8f59e0f889aacdb63c8ffdff3b2">
<div class="sourceCode cell-code" id="cb146"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb146-1"><a href="#cb146-1" aria-hidden="true" tabindex="-1"></a><span class="fu">FeaturePlot</span>(ifnb_sub, <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"CD74"</span>), <span class="at">reduction =</span> <span class="st">'umap.cca'</span>, <span class="at">order =</span> T, <span class="at">slot =</span> <span class="st">'data'</span>, <span class="at">label =</span> <span class="cn">TRUE</span>, <span class="at">repel =</span> <span class="cn">TRUE</span>) <span class="sc">&amp;</span> <span class="fu">scale_color_gradientn</span>(<span class="at">colors =</span> <span class="fu">rev</span>(<span class="fu">brewer.pal</span>(<span class="dv">10</span>, <span class="st">'Spectral'</span>))[<span class="dv">3</span><span class="sc">:</span><span class="dv">8</span>]) <span class="sc">&amp;</span> <span class="fu">labs</span>(<span class="at">color =</span> <span class="st">"log1p</span><span class="sc">\n</span><span class="st">(counts)"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.</code></pre>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="book_files/figure-html/unnamed-chunk-46-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>We can also make a heatmap of the cluster markers:</p>
<div class="cell" data-layout-align="center" data-hash="book_cache/html/unnamed-chunk-47_aa05c30ce998ef5cd66bff52cc156241">
<div class="sourceCode cell-code" id="cb148"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb148-1"><a href="#cb148-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DoHeatmap</span>(<span class="fu">subset</span>(ifnb_sub, <span class="at">downsample =</span> <span class="dv">100</span>), <span class="at">features =</span> top_cluster_markers<span class="sc">$</span>gene, <span class="at">size =</span> <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="book_files/figure-html/unnamed-chunk-47-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>We can customize this heatmap as well:</p>
<div class="cell" data-layout-align="center" data-hash="book_cache/html/unnamed-chunk-48_5cabd1b1d6529ea1cf51d86fd3299c62">
<div class="sourceCode cell-code" id="cb149"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb149-1"><a href="#cb149-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DoHeatmap</span>(<span class="fu">subset</span>(ifnb_sub, <span class="at">downsample =</span> <span class="dv">100</span>), <span class="at">features =</span> top_cluster_markers<span class="sc">$</span>gene, <span class="at">size =</span> <span class="dv">3</span>) <span class="sc">&amp;</span> <span class="fu">scale_fill_viridis</span>() </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for fill is already present.
Adding another scale for fill, which will replace the existing scale.</code></pre>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="book_files/figure-html/unnamed-chunk-48-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>We can adjust which legends are shown, like this:</p>
<div class="cell" data-layout-align="center" data-hash="book_cache/html/unnamed-chunk-49_6ca07517cafa28e793f939186cf49b55">
<div class="sourceCode cell-code" id="cb151"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb151-1"><a href="#cb151-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DoHeatmap</span>(<span class="fu">subset</span>(ifnb_sub, <span class="at">downsample =</span> <span class="dv">100</span>), <span class="at">features =</span> top_cluster_markers<span class="sc">$</span>gene, <span class="at">size =</span> <span class="dv">3</span>) <span class="sc">&amp;</span> <span class="fu">scale_fill_viridis</span>() <span class="sc">&amp;</span> <span class="fu">guides</span>(<span class="at">fill=</span><span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for fill is already present.
Adding another scale for fill, which will replace the existing scale.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: The `&lt;scale&gt;` argument of `guides()` cannot be `FALSE`. Use "none" instead as
of ggplot2 3.3.4.</code></pre>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="book_files/figure-html/unnamed-chunk-49-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Or like this:</p>
<div class="cell" data-layout-align="center" data-hash="book_cache/html/unnamed-chunk-50_55c75171dd0908f045c68a780ca68270">
<div class="sourceCode cell-code" id="cb154"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb154-1"><a href="#cb154-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DoHeatmap</span>(<span class="fu">subset</span>(ifnb_sub, <span class="at">downsample =</span> <span class="dv">100</span>), <span class="at">features =</span> top_cluster_markers<span class="sc">$</span>gene, <span class="at">size =</span> <span class="dv">3</span>) <span class="sc">&amp;</span> <span class="fu">scale_fill_viridis</span>() <span class="sc">&amp;</span>  <span class="fu">guides</span>(<span class="at">colour=</span><span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for fill is already present.
Adding another scale for fill, which will replace the existing scale.</code></pre>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="book_files/figure-html/unnamed-chunk-50-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>How long did it take to run through the notebook?</p>
<div class="cell" data-layout-align="center" data-hash="book_cache/html/unnamed-chunk-51_23a7fe52b4828b110f06d76b0a7dc564">
<div class="sourceCode cell-code" id="cb156"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb156-1"><a href="#cb156-1" aria-hidden="true" tabindex="-1"></a>nb.end.time <span class="ot">&lt;-</span> <span class="fu">Sys.time</span>()</span>
<span id="cb156-2"><a href="#cb156-2" aria-hidden="true" tabindex="-1"></a>nb.end.time <span class="sc">-</span> nb.start.time</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Time difference of 14.16226 mins</code></pre>
</div>
</div>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>